File: backtracking/N&M12.py
import sys
input = sys.stdin.readline

class NandM:
    def __init__(self,n,m,num):
        self.n = n
        self.m = m
        self.num = sorted(num)
        self.arr = [0]*m

    def solve(self, k, st):
        if k==self.m:
            for i in range(self.m): print(self.arr[i], end=" ")
            print()
            return
        
        temp = 0
        for i in range(st, self.n):
            if temp != self.num[i]:
                self.arr[k] = self.num[i]
                temp = self.arr[k]
                self.solve(k+1, i)

def main():
    n,m = map(int, input().split())
    num = list(map(int, input().split()))
    NM = NandM(n,m,num)
    NM.solve(0,0)
    
if __name__ == "__main__":
    main()

File: backtracking/N&M12.cpp
#include<bits/stdc++.h>
using namespace std;

int n,m;
int arr[10];
int num[10];

void solve(int k, int st){
    if(k==m){
        for(int i=0;i<m;i++) cout<<arr[i]<<" ";
        cout<<"\n";
        return;
    }

    int temp = 0;
    for(int i=st;i<n;i++) {
        if(temp != num[i]){
            arr[k] = num[i];
            temp = arr[k];
            solve(k+1,i);
        }
    }
}

int main(void){
    ios::sync_with_stdio(0);
    cin.tie(0);
    cin>>n>>m;
    for(int i=0;i<n;i++) cin>>num[i];
    sort(num, num+n);
    solve(0,0);
}
